backend/src/apis/snapshot_data.rs

```diff
@@ -0,0 +1,286 @@
+        let params = generate_setup_params(K);
+
+        let vk = keygen_vk(&params, &circuit).expect("vk generation should not fail");
+        let pk = keygen_pk(&params, vk.clone(), &circuit).expect("pk generation should not fail");
+
+        let (penultimate_node_left, penultimate_node_right) = &self
+            .mst
+            .penultimate_level_data()
+            .expect("Failed to retrieve penultimate level data");
+
+        let circuit = SolvencyCircuit::<L, N_ASSETS, N_BYTES> {
+            left_node_hash: penultimate_node_left.hash,
+            left_node_balances: penultimate_node_left.balances,
+            right_node_hash: penultimate_node_right.hash,
+            right_node_balances: penultimate_node_right.balances,
+            assets_sum,
+            root_hash: self.mst.root().hash,
+        };
+
+        let instances = circuit.instances();
+
+        self.proof_of_solvency = Some(SolvencyProof::<N_ASSETS> {
+            root_hash: self.mst.root().hash, // equivalant to instances[0]
+            assets_sum,                      // equivalant to instances[1]
+            pk: pk.clone(),
+            proof: full_prover(&params, &pk, circuit.clone(), instances),
+        });
+
+        Ok(())
+    }
+
+    pub fn get_user_proof(&mut self, user_index: u64) -> Result<InclusionProof, &'static str> {
+        let user_proof = self.proofs_of_inclusions.get(&user_index);
+        match user_proof {
+            Some(user_proof) => Ok(InclusionProof {
+                leaf_hash: user_proof.leaf_hash,
+                root_hash: self.mst.root().hash,
+                vk: self.mst_params_and_keys.vk.clone(),
+                pk: self.mst_params_and_keys.pk.clone(),
+                proof: user_proof.proof.clone(),
+            }),
+            None => {
+                let user_proof =
+                    Self::generate_inclusion_proof(&self, user_index as usize).unwrap();
+                self.proofs_of_inclusions
+                    .insert(user_index, user_proof.clone());
+                Ok(InclusionProof {
+                    leaf_hash: user_proof.leaf_hash,
+                    root_hash: self.mst.root().hash,
+                    vk: self.mst_params_and_keys.vk.clone(),
+                    pk: self.mst_params_and_keys.pk.clone(),
+                    proof: user_proof.proof,
+                })
+            }
+        }
+    }
+
+    pub fn get_onchain_proof(&self) -> Result<SolvencyProof<N_ASSETS>, &'static str> {
+        match &self.proof_of_solvency {
+            Some(proof) => Ok(proof.clone()),
+            None => Err("on-chain proof not initialized"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const N_ASSETS: usize = 2;
+    const L: usize = 2 + (N_ASSETS * 2);
+    const LEVELS: usize = 4;
+    const N_BYTES: usize = 31;
+    const K: u32 = 11;
+
+    fn initiate_snapshot_data() -> SnapshotData<LEVELS, L, N_ASSETS, N_BYTES, K> {
```

Comment:
"Initialize" would be a better wording here.

Reply:
Fixed at [771ad22](https://github.com/summa-dev/summa-solvency/pull/88/commits/771ad223da27f5dd9d15e5e3d63d644aae4289f9)